test/TorchSharpTest/TestDisposeScopes.cs

```diff
@@ -0,0 +1,188 @@
+                var am = predictedLabels.argmax(1);
+                var eq = am == labels;
+
+                _testOutputHelper.WriteLine($"Total tensors temporary dispose: {torch.Tensor.TotalCount}"); // Writes 4
+
+                // If the process is long, you can dispose mid stream;
+                d.DisposeEverythingBut(eq);
+
+                // This last value is excluded from the DisposeScope, so it can be used outside.
+                sum = d.Exclude(eq.sum());
+                _testOutputHelper.WriteLine($"Total tensors before dispose: {torch.Tensor.TotalCount}"); // Writes 2
+            }
+
+            _testOutputHelper.WriteLine($"Total tensors after dispose: {torch.Tensor.TotalCount}"); // Writes 1
+            _testOutputHelper.WriteLine($"Tensors disposed: {DisposeScopeManager.DisposedTensorCount}"); // Writes 4
+            _testOutputHelper.WriteLine($"sum:{sum.ToSingle()}");
+        }
+
+        [Fact]
+        public void StackedScopesWorksAsIntended()
+        {
+            torch.Tensor a1;
+            using (var scope1 = torch.NewDisposeScope()) {
+                a1 = 1.ToTensor();
+
+                torch.Tensor a4;
+                using (var scope2 = torch.NewDisposeScope()) {
+                    // Moving to outer scope works
+                    var a2 = 2f.ToTensor();
+                    Assert.Contains(a2, scope2.Disposables);
+                    scope2.Exclude(a2);
+                    Assert.DoesNotContain(a2, scope2.Disposables);
+                    Assert.Contains(a1, scope1.Disposables);
+
+                    // Moving out of the scope alltogether works
+                    var a6 = 6L.ToTensor();
+                    scope2.ExcludeGlobally(a6);
+                    // Assert.DoesNotContain Is broken!
+                    var cont = scope2.Disposables.Contains(a6);
+                    Assert.False(cont);
+                    cont = scope1.Disposables.Contains(a6);
+                    Assert.False(cont);
+
+                    var a3 = 3.ToTensor();
+                    a4 = 4.ToTensor();
+                    scope2.DisposeEverythingBut(a4);
+                    Assert.True(a3.IsDisposed);
+                    Assert.False(a4.IsDisposed);
+
+                    var a5 = 5f.ToTensor();
+                    Assert.Contains(a5, scope2.Disposables);
+                    a5.Dispose();
+                    Assert.DoesNotContain(a5, scope2.Disposables);
+                }
+
+                Assert.False(a1.IsDisposed);
+                Assert.True(a4.IsDisposed);
+            }
+
+            Assert.True(a1.IsDisposed);
+        }
+
+        [Fact]
+        public void TensorsAreDisposedCorrectly()
+        {
+            torch.Tensor data = torch.rand(10, 10);
+            var preTotalCount = torch.Tensor.TotalCount;
+            using (var d = torch.NewDisposeScope()) {
+                var newValue = data * data + data;
+                Assert.True(torch.Tensor.TotalCount > preTotalCount);
```

Comment:
As I mentioned in the issue #450 comments, since xUnit relies on multi-threading, this assertion is likely to fail intermittently. It's too bad, because it's a great way to test that the behavior is what you expect, but I speak from bitter experience.

Reply:
I use the numbers coming from the DisposeScopeManager instead - it's thread static.